src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression?.Parent?.Parent;
+            if (syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            varDeclName = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSitesBuilder = ImmutableDictionary.CreateBuilder<Document, ImmutableArray<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+
+            // Ordering by descending to sort invocations by starting span to account for nested invocations
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            // Adding the original document to account for refactorings that do not have any invocations
+            methodCallSitesBuilder.Add(document, new ArrayBuilder<TInvocationExpressionSyntax>().ToImmutable());
+
+            foreach (var refLocation in referencedLocations)
+            {
+                // Does not support cross-language references currently
+                if (refLocation.Document.Project.Language == document.Project.Language)
+                {
+                    var invocation = refLocation.Location.FindNode(cancellationToken).GetRequiredParent();
+
+                    // Only adding items that are of type InvocationExpressionSyntax
+                    // TODO: in future, add annotation to items that aren't InvocationExpressionSyntax
+                    if (!methodCallSitesBuilder.ContainsKey(refLocation.Document))
+                    {
+                        var invocations = new ArrayBuilder<TInvocationExpressionSyntax>();
+                        if (syntaxFacts.IsInvocationExpression(invocation))
```

Comment:
1. lift the IsInvocationExpression to be your first check.  
2. no need for syntax facts check.  just do `if (reference is not TInvocationExpressionSyntax invocation) continue;`

Reply:
added